lib/rspec/core/backtrace_formatter.rb

```diff
@@ -34,20 +37,41 @@ def format_backtrace(backtrace, options={})
         return [] unless backtrace
         return backtrace if options[:full_backtrace] || backtrace.empty?
 
-        backtrace.map { |l| backtrace_line(l) }.compact.
-          tap do |filtered|
-            if filtered.empty?
-              filtered.concat backtrace
-              filtered << ""
-              filtered << "  Showing full backtrace because every line was filtered out."
-              filtered << "  See docs for RSpec::Configuration#backtrace_exclusion_patterns and"
-              filtered << "  RSpec::Configuration#backtrace_inclusion_patterns for more information."
+        prelines = []
+        has_matched = false
+
+        filtered = backtrace.map do |line|
+          if exclude?(line)
+            unless has_matched
+              if !preexclude?(line)
+                prelines << raw_backtrace_line(line)
+              else
+                prelines = []
+                has_matched = true
+              end
+              nil
             end
+          else
+            has_matched = true
+            backtrace_line(line)
           end
+        end.compact
+
+        filtered = prelines + filtered if has_matched
+
+        if filtered.empty?
+          filtered.concat backtrace
+          filtered << ""
+          filtered << "  Showing full backtrace because every line was filtered out."
+          filtered << "  See docs for RSpec::Configuration#backtrace_exclusion_patterns and"
+          filtered << "  RSpec::Configuration#backtrace_inclusion_patterns for more information."
+        end
+
+        filtered
       end
```

Comment:
The `tap` can still be used here, refactor into a pipeline to enumerate and select/reject based on your new critera.